File: src/main/java/com/now/nowbot/model/beatmapParse/OsuFile.java
package com.now.nowbot.model.beatmapParse;

import com.now.nowbot.entity.BeatMapFileLite;
import com.now.nowbot.entity.BeatmapLite;
import com.now.nowbot.model.beatmapParse.parse.*;
import com.now.nowbot.model.enums.OsuMode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class OsuFile {
    static final Logger         log =  LoggerFactory.getLogger(OsuFile.class.getName());
    private      BeatmapGeneral general;

    private BufferedReader reader;

    public static OsuFile getInstance(String osuFileStr) throws IOException {
        return new OsuFile(new BufferedReader(
                new InputStreamReader(
                        new ByteArrayInputStream(
                                osuFileStr.getBytes(StandardCharsets.UTF_8)
                        )
                )
        ));
    }

    public static OsuFile getInstance(BufferedReader read) throws IOException {
        return new OsuFile(read);
    }

    public static BeatMapFileLite parseInfo(BufferedReader read) throws IOException {
        var versionStr = read.readLine();
        if (versionStr == null || !versionStr.startsWith("osu file format v")) {
            log.error("解析错误,文件无效 第一行为:[{}]", versionStr);
            throw new RuntimeException("解析错误,文件无效");
        }
        var bf = new BeatMapFileLite();
        HashMap<String, String> info = new HashMap<>(5);
        info.put("AudioFilename", null);
        info.put("Mode", null);
        info.put("BeatmapID", null);
        info.put("BeatmapSetID", null);
        String line;
        // 逐行
        while ((line = read.readLine()) != null) {
            if (line.startsWith("[General]") || line.startsWith("[Metadata]")) {
                // 读取 General 块
                parseAny(read, info);
            } if (line.startsWith("[Events]")) {
                break;
            }
        }

        while ((line = read.readLine()) != null) {
            if (line.startsWith("//") || line.isBlank()) {
                continue;
            } else if (line.startsWith("[")) {
                break;
            }

            int start = line.indexOf('"');
            int end = line.lastIndexOf('"');

            bf.setBackground(line.substring(start + 1, end));
            break;
        }

        if (info.get("AudioFilename") != null) {
            bf.setAudio(info.get("AudioFilename"));
        }
        if (info.get("Mode") != null) {
            bf.setMode(Integer.parseInt(info.get("Mode")));
        }
        if (info.get("BeatmapID") != null) {
            bf.setBid(Long.parseLong(info.get("BeatmapID")));
        }
        if (info.get("BeatmapSetID") != null) {
            bf.setSid(Long.parseLong(info.get("BeatmapSetID")));
        }

        read.close();
        return bf;
    }

    public OsuBeatmapAttributes getOsu() throws IOException {
        switch (getMode()) {
            case OSU -> {
                return new OsuBeatmapAttributes(reader, general);
            }
            case TAIKO -> {
                return new TaikoBeatmapAttributes(reader, general);
            }
            case CATCH -> {
                return new CatchBeatmapAttributes(reader, general);
            }
            case MANIA -> {
                return new ManiaBeatmapAttributes(reader, general);
            }
        }
        throw new RuntimeException("mode type error");
    }


    public ManiaBeatmapAttributes getCatch() throws IOException {
        if (getMode() != OsuMode.CATCH && getMode() != OsuMode.OSU) throw new RuntimeException("mode error");
        return new ManiaBeatmapAttributes(reader, general);
    }

    public ManiaBeatmapAttributes getTaiko() throws IOException {
        if (getMode() != OsuMode.TAIKO && getMode() != OsuMode.OSU) throw new RuntimeException("mode error");
        return new ManiaBeatmapAttributes(reader, general);
    }

    public ManiaBeatmapAttributes getMania() throws IOException {
        if (getMode() != OsuMode.MANIA && getMode() != OsuMode.OSU) throw new RuntimeException("mode error");
        return new ManiaBeatmapAttributes(reader, general);
    }

    OsuFile(String osuFileStr) throws IOException {
//        转化为 BufferedReader 逐行读取
        new OsuFile(new BufferedReader(
                new InputStreamReader(
                        new ByteArrayInputStream(
                                osuFileStr.getBytes(StandardCharsets.UTF_8)
                        )
                )
        ));
    }

    /**
     * 逐行读取
     *
     * @param read osu file
     * @throws {@link IOException} io exception
     */
    @SuppressWarnings("")
    OsuFile(BufferedReader read) throws IOException {
        // 这里将 version 与 general 读取拆分出来, 然后解析出的所有属性放入 Attributes 类中
        // 一是在读取的时候可以自动判断并生成对应的类型, 防止读取未知类型谱面的时候错过其他类型的解析, 比如骂娘的 Column
        // 另外也能将 osu 转谱其他模式
        // 读取 version
        var versionStr = read.readLine();
        int versionInt;
        if (versionStr != null && versionStr.startsWith("osu file format v")) {
            versionInt = Integer.parseInt(versionStr.substring(17));
        } else {
            throw new RuntimeException("解析错误,文件无效");
        }
        this.general = new BeatmapGeneral(versionInt);
        this.reader = read;
        String line;
        // 逐行
        while ((line = read.readLine()) != null && line.equals("")) ;
        if (line != null && line.startsWith("[General]")) {
            // 读取 General 块
            parseGeneral(read);
        } else {
            throw new RuntimeException("解析错误,缺失 [General] 块");
        }
    }


    void parseGeneral(BufferedReader reader) throws IOException {
        String line;
        while ((line = reader.readLine()) != null && !line.equals("")) {
            var entity = line.split(":");
            if (entity.length != 2) {
                continue;
            }
            var key = entity[0].trim();
            var val = entity[1].trim();

            switch (key) {
                case "Mode" -> general.setMode(OsuMode.getMode(val));
                case "StackLeniency" -> general.setStackLeniency(Double.parseDouble(val));
                case "SampleSet" -> general.setSampleSet(val);
            }
        }
    }

    private static void parseAny(BufferedReader reader, Map<String, String> parseMap) throws IOException {
        String line;
        while ((line = reader.readLine()) != null && !line.equals("")) {
            var entity = line.split(":");
            if (entity.length != 2) {
                continue;
            }
            var key = entity[0].trim();
            var val = entity[1].trim();

            if (parseMap != null) {
                if (parseMap.containsKey(key)){
                    parseMap.put(key, val);
                }
            }
        }
    }

    public OsuMode getMode() {
        return general.getMode();
    }
}

File: src/main/java/com/now/nowbot/model/beatmapParse/OsuFile.kt
package com.now.nowbot.model.beatmapParse

import com.now.nowbot.entity.BeatmapFileLite
import com.now.nowbot.model.beatmapParse.parse.*
import com.now.nowbot.model.enums.OsuMode
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.BufferedReader
import java.io.IOException

class OsuFile @Throws(IOException::class) constructor(private val reader: BufferedReader) {

    private val general: BeatmapGeneral

    val mode: OsuMode
        get() = general.mode ?: OsuMode.OSU

    init {
        val versionStr = reader.readLine()
        // 修正：确保 versionStr 不为 null 且正确 trim
        val versionInt = if (versionStr != null && versionStr.trim().startsWith("osu file format v")) {
            versionStr.substring(17).trim().toInt()
        } else {
            throw RuntimeException("解析错误,文件无效")
        }

        this.general = BeatmapGeneral(versionInt)

        var line: String?
        // 修正：安全跳过空白行直到找到 [General]
        while (true) {
            line = reader.readLine()
            if (line == null) throw RuntimeException("解析错误,缺失 [General] 块")
            if (line.isNotBlank()) break
        }

        if (line.startsWith("[General]")) {
            parseGeneral(reader)
        } else {
            throw RuntimeException("解析错误,缺失 [General] 块")
        }
    }

    @Throws(IOException::class)
    private fun parseGeneral(reader: BufferedReader) {
        while (true) {
            val line = reader.readLine()?.takeIf { it.isNotBlank() } ?: break
            val entity = line.split(":", limit = 2)
            if (entity.size != 2) continue

            val key = entity[0].trim()
            val value = entity[1].trim()

            when (key) {
                "Mode" -> general.mode = OsuMode.getMode(value)
                "StackLeniency" -> general.stackLeniency = value.toDoubleOrNull() ?: 0.0
                "SampleSet" -> general.sampleSet = value
            }
        }
    }

    @Throws(IOException::class)
    fun getOsu(): OsuBeatmapAttributes {
        return when (mode) {
            OsuMode.OSU -> OsuBeatmapAttributes(reader, general)
            OsuMode.TAIKO -> TaikoBeatmapAttributes(reader, general)
            OsuMode.CATCH -> CatchBeatmapAttributes(reader, general)
            OsuMode.MANIA -> ManiaBeatmapAttributes(reader, general)
            else -> throw RuntimeException("mode type error")
        }
    }

    @Throws(IOException::class)
    fun getCatch(): ManiaBeatmapAttributes {
        if (mode != OsuMode.CATCH && mode != OsuMode.OSU) throw RuntimeException("mode error")
        return ManiaBeatmapAttributes(reader, general)
    }

    @Throws(IOException::class)
    fun getTaiko(): ManiaBeatmapAttributes {
        if (mode != OsuMode.TAIKO && mode != OsuMode.OSU) throw RuntimeException("mode error")
        return ManiaBeatmapAttributes(reader, general)
    }

    @Throws(IOException::class)
    fun getMania(): ManiaBeatmapAttributes {
        if (mode != OsuMode.MANIA && mode != OsuMode.OSU) throw RuntimeException("mode error")
        return ManiaBeatmapAttributes(reader, general)
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(OsuFile::class.java.name)

        /**
         * 对应 Java 的 OsuFile.getInstance(String osuFileStr)
         */
        @JvmStatic
        @Throws(IOException::class)
        fun getInstance(osuFileStr: String): OsuFile {
            val reader = BufferedReader(
                java.io.InputStreamReader(
                    java.io.ByteArrayInputStream(osuFileStr.toByteArray(java.nio.charset.StandardCharsets.UTF_8))
                )
            )
            return OsuFile(reader)
        }

        /**
         * 对应 Java 的 OsuFile.getInstance(BufferedReader read)
         */
        @JvmStatic
        @Throws(IOException::class)
        fun getInstance(read: BufferedReader): OsuFile {
            return OsuFile(read)
        }


        @Throws(IOException::class)
        fun parseInfo(reader: BufferedReader): BeatmapFileLite {

            return reader.use { r ->
                val versionStr = r.readLine()
                if (versionStr == null || !versionStr.trim().startsWith("osu file format v")) {
                    log.error("解析错误,文件无效 第一行为:{}", versionStr)
                    throw RuntimeException("解析错误,文件无效")
                }
                if (versionStr.endsWith("v3")) {
                    throw RuntimeException("不支持v3版本的解析")
                }

                val bf = BeatmapFileLite()
                val info = mutableMapOf<String, String?>(
                    "AudioFilename" to null,
                    "Mode" to null,
                    "BeatmapID" to null,
                    "BeatmapSetID" to null
                )

                var line: String?
                // 修正：模仿 Java 的 while ((line = read.readLine()) != null)
                while (reader.readLine().also { line = it } != null) {
                    val currentLine = line!!
                    if (currentLine.startsWith("[General]") || currentLine.startsWith("[Metadata]")) {
                        parseAny(reader, info)
                    }
                    if (currentLine.startsWith("[Events]")) {
                        break
                    }
                }

                // 解析 Background 逻辑
                while (reader.readLine().also { line = it } != null) {
                    val currentLine = line!!
                    if (currentLine.isBlank() || currentLine.startsWith("//")) continue
                    if (currentLine.startsWith("[")) break

                    val lineSplit = currentLine.split(",")
                    if (lineSplit.size >= 3 && lineSplit[0] == "0" && lineSplit[1] == "0" &&
                        lineSplit[2].startsWith("\"") && lineSplit[2].endsWith("\"")
                    ) {

                        val bgStr = lineSplit[2]
                        bf.background = bgStr.substring(1, bgStr.length - 1)
                        break
                    }
                }

                info["AudioFilename"]?.let { bf.audio = it }
                info["Mode"]?.let { bf.modeInt = it.toInt() }
                info["BeatmapID"]?.let { bf.bid = it.toLong() }
                info["BeatmapSetID"]?.let { bf.sid = it.toLong() }

                return@use bf
            }
        }

        @Throws(IOException::class)
        private fun parseAny(reader: BufferedReader, parseMap: MutableMap<String, String?>) {
            while (true) {
                val line = reader.readLine()?.takeIf { it.isNotBlank() } ?: break
                val entity = line.split(":".toRegex(), limit = 2)
                if (entity.size != 2) continue

                val key = entity[0].trim()
                if (parseMap.containsKey(key)) {
                    parseMap[key] = entity[1].trim()
                }
            }
        }
    }
}